File: VB/HighlightChangedCellBehavior.vb
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Grid
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows
Imports System.Windows.Media

Namespace HighlightModifiedCells

    Public Class ChangedCellsHighlightBehavior
        Inherits DevExpress.Mvvm.UI.Interactivity.Behavior(Of DevExpress.Xpf.Grid.GridControl)

        Const UnboundColumnPrefix As String = "IsEdited_"

        Private modifiedCells As System.Collections.Generic.Dictionary(Of System.Tuple(Of Object, String), Boolean) = New System.Collections.Generic.Dictionary(Of System.Tuple(Of Object, String), Boolean)()

        Private originalValues As System.Collections.Generic.Dictionary(Of System.Tuple(Of Object, String), Object) = New System.Collections.Generic.Dictionary(Of System.Tuple(Of Object, String), Object)()

        Public Shared ReadOnly HighlightBrushProperty As System.Windows.DependencyProperty = System.Windows.DependencyProperty.Register("HighlightBrush", GetType(System.Windows.Media.Brush), GetType(HighlightModifiedCells.ChangedCellsHighlightBehavior))

        Public Property HighlightBrush As Brush
            Get
                Return CType(Me.GetValue(HighlightModifiedCells.ChangedCellsHighlightBehavior.HighlightBrushProperty), System.Windows.Media.Brush)
            End Get

            Set(ByVal value As Brush)
                Me.SetValue(HighlightModifiedCells.ChangedCellsHighlightBehavior.HighlightBrushProperty, value)
            End Set
        End Property

        Protected ReadOnly Property View As TableView
            Get
                Return CType(Me.AssociatedObject.View, DevExpress.Xpf.Grid.TableView)
            End Get
        End Property

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            AddHandler Me.AssociatedObject.CustomUnboundColumnData, AddressOf Me.OnGridCustomUnboundColumnData
            AddHandler Me.AssociatedObject.Loaded, AddressOf Me.OnGridLoaded
        End Sub

        Protected Overrides Sub OnDetaching()
            RemoveHandler Me.AssociatedObject.CustomUnboundColumnData, AddressOf Me.OnGridCustomUnboundColumnData
            RemoveHandler Me.View.CellValueChanged, AddressOf Me.OnViewCellValueChanged
            MyBase.OnDetaching()
        End Sub

        Private Sub OnGridLoaded(ByVal sender As Object, ByVal e As System.Windows.RoutedEventArgs)
            RemoveHandler Me.AssociatedObject.Loaded, AddressOf Me.OnGridLoaded
            AddHandler Me.View.CellValueChanged, AddressOf Me.OnViewCellValueChanged
            Me.CreateColumnsAndConditions()
        End Sub

        Private Sub OnViewCellValueChanged(ByVal sender As Object, ByVal e As DevExpress.Xpf.Grid.CellValueChangedEventArgs)
            If Not HighlightModifiedCells.ChangedCellsHighlightBehavior.IsServiceColumn(e.Column) Then
                Dim fieldName = $"{HighlightModifiedCells.ChangedCellsHighlightBehavior.UnboundColumnPrefix} {e.Column.FieldName}"
                Dim key = System.Tuple.Create(e.Row, fieldName)
                Dim originalValue As Object
                Dim isModified As Boolean = True
                If Me.originalValues.TryGetValue(key, originalValue) Then
                    isModified = Not System.[Object].Equals(originalValue, e.Value)
                Else
                    Me.originalValues(key) = e.OldValue
                End If

                Me.AssociatedObject.SetCellValue(e.RowHandle, fieldName, isModified)
            End If
        End Sub

        Private Sub OnGridCustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.Xpf.Grid.GridColumnDataEventArgs)
            If HighlightModifiedCells.ChangedCellsHighlightBehavior.IsServiceColumn(e.Column) Then
                Dim key = System.Tuple.Create(Me.AssociatedObject.GetRowByListIndex(e.ListSourceRowIndex), e.Column.FieldName)
                If e.IsGetData Then
                    Dim res As Boolean
                    e.Value = If(Me.modifiedCells.TryGetValue(key, res), res, False)
                End If

                If e.IsSetData Then Me.modifiedCells(key) = CBool(e.Value)
            End If
        End Sub

        Private Shared Function IsServiceColumn(ByVal column As DevExpress.Xpf.Grid.GridColumn) As Boolean
            Return column.FieldName.StartsWith(HighlightModifiedCells.ChangedCellsHighlightBehavior.UnboundColumnPrefix)
        End Function

        Private Sub CreateColumnsAndConditions()
            For Each column As DevExpress.Xpf.Grid.GridColumn In Me.AssociatedObject.Columns.ToList()
                Dim unboundColumn As DevExpress.Xpf.Grid.GridColumn = New DevExpress.Xpf.Grid.GridColumn()
                unboundColumn.FieldName = HighlightModifiedCells.ChangedCellsHighlightBehavior.UnboundColumnPrefix & column.FieldName
                unboundColumn.UnboundDataType = GetType(Boolean)
                unboundColumn.Visible = False
                unboundColumn.ShowInColumnChooser = False
                Me.AssociatedObject.Columns.Add(unboundColumn)
                Me.View.FormatConditions.Add(New DevExpress.Xpf.Grid.FormatCondition() With {.FieldName = column.FieldName, .Expression = String.Format("[{0}] = true", unboundColumn.FieldName), .Format = New DevExpress.Xpf.Core.ConditionalFormatting.Format() With {.Background = Me.HighlightBrush}})
            Next
        End Sub
    End Class
End Namespace

File: CS/HighlightChangedCellBehavior.cs
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Grid;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Windows.Media;

namespace HighlightModifiedCells {
    public class ChangedCellsHighlightBehavior : Behavior<GridControl> {
        const string UnboundColumnPrefix = "IsEdited_";
        Dictionary<Tuple<object, string>, bool> modifiedCells = new Dictionary<Tuple<object, string>, bool>();
        Dictionary<Tuple<object, string>, object> originalValues = new Dictionary<Tuple<object, string>, object>();
        public static readonly DependencyProperty HighlightBrushProperty = DependencyProperty.Register("HighlightBrush", typeof(Brush), typeof(ChangedCellsHighlightBehavior));
        public Brush HighlightBrush {
            get { return (Brush)GetValue(HighlightBrushProperty); }
            set { SetValue(HighlightBrushProperty, value); }
        }
        protected TableView View {
            get { return (TableView)AssociatedObject.View; }
        }
        protected override void OnAttached() {
            base.OnAttached();
            AssociatedObject.CustomUnboundColumnData += OnGridCustomUnboundColumnData;
            AssociatedObject.Loaded += OnGridLoaded;
        }
        protected override void OnDetaching() {
            AssociatedObject.CustomUnboundColumnData -= OnGridCustomUnboundColumnData;
            View.CellValueChanged -= OnViewCellValueChanged;
            base.OnDetaching();
        }
        void OnGridLoaded(object sender, System.Windows.RoutedEventArgs e) {
            AssociatedObject.Loaded -= OnGridLoaded;
            View.CellValueChanged += OnViewCellValueChanged;
            CreateColumnsAndConditions();
        }
        void OnViewCellValueChanged(object sender, CellValueChangedEventArgs e) {
            if (!IsServiceColumn(e.Column)) {
                var fieldName = $"{UnboundColumnPrefix}{e.Column.FieldName}";
                var key = Tuple.Create(e.Row, fieldName);
                object originalValue;
                bool isModified = true;
                if (originalValues.TryGetValue(key, out originalValue))
                    isModified = !Equals(originalValue, e.Value);
                else
                    originalValues[key] = e.OldValue;
                AssociatedObject.SetCellValue(e.RowHandle, fieldName, isModified);
            }
        }
        void OnGridCustomUnboundColumnData(object sender, GridColumnDataEventArgs e) {
            if (IsServiceColumn(e.Column)) {
                var key = Tuple.Create(AssociatedObject.GetRowByListIndex(e.ListSourceRowIndex), e.Column.FieldName);
                if (e.IsGetData) {
                    bool res;
                    e.Value = modifiedCells.TryGetValue(key, out res) ? res : false;
                }
                if (e.IsSetData)
                    modifiedCells[key] = (bool)e.Value;
            }
        }
        static bool IsServiceColumn(GridColumn column) {
            return column.FieldName.StartsWith(UnboundColumnPrefix);
        }
        void CreateColumnsAndConditions() {
            foreach (GridColumn column in AssociatedObject.Columns.ToList()) {
                GridColumn unboundColumn = new GridColumn();
                unboundColumn.FieldName = UnboundColumnPrefix + column.FieldName;
                unboundColumn.UnboundDataType = typeof(bool);
                unboundColumn.Visible = false;
                unboundColumn.ShowInColumnChooser = false;
                AssociatedObject.Columns.Add(unboundColumn);
                View.FormatConditions.Add(new FormatCondition() {
                    FieldName = column.FieldName,
                    Expression = string.Format("[{0}] = true", unboundColumn.FieldName),
                    Format = new DevExpress.Xpf.Core.ConditionalFormatting.Format() { Background = HighlightBrush }
                });
            }
        }
    }
}
